Arduboy core class (that other classes build upon)

joshgoebel opened the pull request:
This is a common pattern (seen in Gamebuino and other libs).  The problem is how to have very different "modes" (buffer, greyscale, tiled, etc.) but a single core lib.  Or even to answer the question of some people who want a terribly slim library vs those that want a LOT more functionality.  The core library can support both ways of thinking.  The answer is create a logical class hierarchy (not even a deep one).  I'd like to propose something like this:

(Print is included below to allow printing to screen, which comes from the Print.cpp Arduino core library)
- ArduboyCore (low-level hardware boot, button polling, power saving, safe mode, etc)
- Arduboy (ArduboyCore, Print, Buffered graphics) - same as today
- ArduboyText (ArduboyCore, Print, buffered text array)
- ArduboyTiled (ArduboyCore, support for array of tiles and buffered scrollable display rendered with tiles)

So pretty much just the core and then the classes MOST people would be using lay in on top of the core to provide the drawing modes, etc.  I'm not even saying all these classes need to be in core... it could be that ArduboyText and ArduboyTiled are developed as separate libraries and maintained in their own right.  Splitting the core hardware stuff into ArduboyCore allows people who want to build these libraries to still take advantage of the core lib, even if they're wanting to do something far different than a buffer backed display.  They still get the same boot up code, power saving, etc...

This encourages people to use the core lib (even if only for ArduboyCore) vs forking it into 5 different variants for different types of displays that do not want a buffer (or want a differently configured buffer).  

This pull request is the beginning of this effort.  It starts by creating ArduboyCore and moving all the hardware level stuff there.  Arduboy becomes a subclass of ArduboyCore and has almost exactly the same API as before.  I change `drawScreen` to `paintScreen` for clarify, but I could add a proxy `drawScreen` back to the `Arduboy` class if that would be helpful.

joshgoebel committed:
move LCD boot command sequence into PROGMEM

joshgoebel committed:
split out core Arduboy hardware stuff

yyyc514 commented:
This also makes it easy to say reuse things like Adafruit_GFX without having to completely re-invent the wheel (unless we really, really want to).  Unless we're going to hand tune ALL those functions, why not stand on the shoulders of those who have come before.  Of course I'd optimize drawPixel and drawBitmap as much as you could even if you subclasses GFX, but mostly you could leave all the other methods alone and then our core library is a lot slimmer than before.  

It would look like this:

```
class ArduboyGFX: public Adafruit_GFX
{
void drawPixel(int16_t x, int16_t y, uint16_t color);

private:
unsigned char screenBuffer[(HEIGHT*WIDTH)/8];
}
```

Just implement drawPixel (that talks to screenBuffer) and then you're using Arduboy with the Adafruit_GFX library.  If they add a really cool graphics function, you get it for free - we don't have to worry about porting it over to Arduboy lib.

joshgoebel committed:
hand optimized paintScreen from memory buffer

- SPI /4 SPI.transfer(): 3700 micros
- SPI /2 SPI.transfer(): 2670 micros
- SPI /2 hand-optimized loop (tight as it can get): 1170 micros

joshgoebel committed:
add comments for button polling

joshgoebel committed:
fully document ArduboyCore

yyyc514 commented:
Ping.

joshgoebel committed:
bump SPI clock to /2 because assembly expects that speed
